nipype/pipeline/engine/workflows.py

```diff
@@ -903,21 +913,32 @@ def _set_node_input(self, node, param, source, sourceinfo):
         node.set_input(param, deepcopy(newval))
 
     def _get_all_nodes(self):
-        allnodes = []
-        for node in self._graph.nodes():
-            if isinstance(node, Workflow):
-                allnodes.extend(node._get_all_nodes())
-            else:
-                allnodes.append(node)
+        allnodes = [
+            *self._nodes_cache.difference(self._nested_workflows_cache)
+        ]  # all nodes that are not workflows
+        for node in self._nested_workflows_cache:
+            allnodes.extend(node._get_all_nodes())
         return allnodes
 
+    def _update_node_cache(self):
+        nodes = set(self._graph)
+
+        added_nodes = nodes.difference(self._nodes_cache)
+        removed_nodes = self._nodes_cache.difference(nodes)
+
+        self._nodes_cache = nodes
+        self._nested_workflows_cache.difference_update(removed_nodes)
+
+        for node in added_nodes:
+            if isinstance(node, Workflow):
+                self._nested_workflows_cache.add(node)
+
     def _has_node(self, wanted_node):
-        for node in self._graph.nodes():
-            if wanted_node == node:
+        if wanted_node in self._nodes_cache:
+            return True
+        for node in self._nested_workflows_cache:
+            if node._has_node(wanted_node):
                 return True
-            if isinstance(node, Workflow):
-                if node._has_node(wanted_node):
-                    return True
         return False
```

Comment:
Small style thought:

```Python
def _has_node(self, wanted_node):
    return (
        wanted_node in self._nodes_cache or
        any(wf._has_node(wanted_node) for wf in self._nested_workflows_cache)
    )
```

This feels more descriptive than procedural but does the same comparisons in the same order, with short-circuiting. Feel free to reject.

Reply:
I like this. GitHub doesn't let me commit this as a suggestion, so I will need to pull it out and commit it manually :-)